sdk/storage/Azure.Storage.Common/src/Sas/SasQueryParameters.cs

```diff
@@ -0,0 +1,437 @@
+    /// </summary>
+    public partial class SasQueryParameters
+    {
+        /// <summary>
+        /// The default service version to use for Shared Access Signatures.
+        /// </summary>
+        public const string DefaultSasVersion = Constants.DefaultSasVersion;
+
+        // SASTimeFormat represents the format of a SAS start or expiry time. Use it when formatting/parsing a time.Time.
+        // ISO 8601 uses "yyyy'-'MM'-'dd'T'HH':'mm':'ss"
+        const string TimeFormat = "yyyy-MM-ddTHH:mm:ssZ";
+
+        /// <summary>
+        /// FormatTimesForSASSigning converts a time.Time to a snapshotTimeFormat string suitable for a
+        /// SASField's StartTime or ExpiryTime fields. Returns "" if value.IsZero().
+        /// </summary>
+        /// <param name="time"></param>
+        /// <returns></returns>
+        internal static string FormatTimesForSasSigning(DateTimeOffset time) =>
+            // "yyyy-MM-ddTHH:mm:ssZ"
+            (time == new DateTimeOffset()) ? "" : time.ToString(TimeFormat, CultureInfo.InvariantCulture);
+
+        // All members are immutable or values so copies of this struct are thread safe.
+
+        // sv
+        readonly string version;
+
+        // ss
+        readonly string services;
+
+        // srt
+        readonly string resourceTypes;
+
+        // Use auto property
+#pragma warning disable IDE0032
+
+        // spr
+        readonly SasProtocol protocol;
+
+        // st
+        readonly DateTimeOffset startTime;
+
+        // se
+        readonly DateTimeOffset expiryTime;
+
+        // sip
+        readonly IPRange ipRange;
+
+        // Use auto property
+#pragma warning restore IDE0032 
+
+        // si
+        readonly string identifier;
+
+        // sr
+        readonly string resource;
+
+        // sp
+        readonly string permissions;
+
+        // sig
+        readonly string signature;
+
+        /// <summary>
+        /// Gets the storage service version to use to authenticate requests
+        /// made with this shared access signature, and the service version to
+        /// use when handling requests made with this shared access signature.
+        /// </summary>
+        public string Version => this.version ?? DefaultSasVersion;
+
+        /// <summary>
+        /// Gets the signed services accessible with an account level shared
+        /// access signature. 
+        /// </summary>
+        public string Services => this.services ?? String.Empty;
+
+        /// <summary>
+        /// Gets which resources are accessible via the shared access signature.
+        /// </summary>
+        public string ResourceTypes => this.resourceTypes ?? String.Empty;
+
+        /// <summary>
+        /// Optional. Specifies the protocol permitted for a request made with
+        /// the shared access signature.
+        /// </summary>
+        public SasProtocol Protocol => this.protocol;
```

Comment:
Jeff has a big preference for `readonly` because you can't ever get it wrong when making future updates to the type without explicitly deleting the modifier.

Reply:
We might push back on that.  I'd prefer to rely on the compiler, rather than adding complexity to the code.

Reply:
This discussion can be pended